File: android/Streamo/app/src/main/java/com/streamo/app/util/TVLogic.kt
package com.streamo.app.util

import com.streamo.app.data.remote.dto.TmdbEpisodeDetail
import com.streamo.app.data.remote.dto.TmdbItem
import java.util.Calendar

object TVLogic {
    const val WATCHED_THRESHOLD = 0.9

    fun isFutureDate(dateStr: String?): Boolean {
        val parts = ymd(dateStr) ?: return false
        val cal = Calendar.getInstance()
        val today = cal.apply { set(Calendar.HOUR_OF_DAY, 0); set(Calendar.MINUTE, 0); set(Calendar.SECOND, 0); set(Calendar.MILLISECOND, 0) }.time
        val date = cal.apply { set(parts.first, parts.second - 1, parts.third, 0, 0, 0); set(Calendar.MILLISECOND, 0) }.time
        return date > today
    }

    private fun ymd(s: String?): Triple<Int, Int, Int>? {
        if (s.isNullOrEmpty() || s.length < 10) return null
        val parts = s.substring(0, 10).split("-")
        if (parts.size != 3) return null
        val y = parts[0].toIntOrNull() ?: return null
        val m = parts[1].toIntOrNull() ?: return null
        val d = parts[2].toIntOrNull() ?: return null
        return Triple(y, m, d)
    }

    fun countEpisodesUpTo(item: TmdbItem, season: Int?, episode: Int?): Int {
        if (season == null || episode == null) return item.numberOfEpisodes ?: 0
        var count = 0
        for (s in item.seasons.orEmpty().filter { it.seasonNumber != 0 }) {
            when {
                s.seasonNumber < season -> count += s.episodeCount ?: 0
                s.seasonNumber == season -> count += episode
            }
        }
        return count
    }

    fun effectiveLastEpisode(item: TmdbItem): Pair<Int, Int>? {
        val nea = item.nextEpisodeToAir
        if (nea?.seasonNumber != null && nea.episodeNumber != null && nea.airDate != null && !isFutureDate(nea.airDate)) {
            return Pair(nea.seasonNumber, nea.episodeNumber)
        }
        val lea = item.lastEpisodeToAir
        if (lea?.seasonNumber != null && lea.episodeNumber != null) {
            return Pair(lea.seasonNumber, lea.episodeNumber)
        }
        return null
    }

    fun airedEpisodesCount(item: TmdbItem): Int {
        val lea = effectiveLastEpisode(item) ?: return 0
        return countEpisodesUpTo(item, lea.first, lea.second)
    }

    fun airedEpisodesInSeason(item: TmdbItem, season: Int): Int {
        val info = item.seasons?.find { it.seasonNumber == season } ?: return 0
        val total = info.episodeCount ?: 0
        val lea = effectiveLastEpisode(item) ?: return total
        return when {
            season < lea.first -> total
            season > lea.first -> 0
            else -> minOf(total, lea.second)
        }
    }

    fun episodesBefore(item: TmdbItem, season: Int, episode: Int): Int {
        if (season <= 0) return 0
        return countEpisodesUpTo(item, season, maxOf(0, episode - 1))
    }

    fun availableSeasons(item: TmdbItem): List<Int> {
        val seasons = item.seasons.orEmpty().filter { it.seasonNumber > 0 }
        val nums: List<Int> = if (effectiveLastEpisode(item) != null) {
            val lastAired = effectiveLastEpisode(item)!!.first
            seasons.filter { it.seasonNumber <= lastAired }.map { it.seasonNumber }
        } else {
            seasons.filter { it.airDate != null && !isFutureDate(it.airDate) }.map { it.seasonNumber }
        }
        return nums.sorted().ifEmpty { listOf(1) }
    }

    fun nextEpisode(item: TmdbItem, season: Int, episode: Int): Pair<Int, Int>? {
        val currentAired = airedEpisodesInSeason(item, season)
        if (currentAired > 0 && episode + 1 <= currentAired) {
            return Pair(season, episode + 1)
        }
        val future = item.seasons.orEmpty()
            .filter { it.seasonNumber > season && airedEpisodesInSeason(item, it.seasonNumber) > 0 }
            .sortedBy { it.seasonNumber }
            .firstOrNull()
        return future?.let { Pair(it.seasonNumber, 1) }
    }

    /**
     * Mirror of [nextEpisode]: the previous episode's (season, episode) coordinate, or null
     * if [season]/[episode] is already the earliest aired episode (S1E1 in practice).
     */
    fun previousEpisode(item: TmdbItem, season: Int, episode: Int): Pair<Int, Int>? {
        if (episode > 1) {
            return Pair(season, episode - 1)
        }
        val past = item.seasons.orEmpty()
            .filter { it.seasonNumber < season && airedEpisodesInSeason(item, it.seasonNumber) > 0 }
            .sortedByDescending { it.seasonNumber }
            .firstOrNull()
        return past?.let { Pair(it.seasonNumber, airedEpisodesInSeason(item, it.seasonNumber)) }
    }

    fun airedEpisodeList(episodes: List<TmdbEpisodeDetail>, item: TmdbItem, season: Int): List<TmdbEpisodeDetail> {
        val sorted = episodes.sortedBy { it.episodeNumber }
        val lea = effectiveLastEpisode(item)
        return if (lea != null) {
            when {
                season < lea.first -> sorted
                season > lea.first -> emptyList()
                else -> sorted.filter { it.episodeNumber <= lea.second }
            }
        } else {
            sorted.filter { !isFutureDate(it.airDate) && it.airDate != null }
        }
    }
}

File: vidaaos/src/util/tvlogic.ts
// Port of TVLogic.kt. Season/episode traversal + aired-episode gating.
// TmdbItem/TmdbSeasonInfo/TmdbEpisodeRef/TmdbEpisodeDetail are minimal structural
// mirrors of the Kotlin DTOs in data/remote/dto/TmdbModels.kt — only the fields
// TVLogic actually reads are required.

export const WATCHED_THRESHOLD = 0.9;

export interface TmdbEpisodeRef {
  seasonNumber: number | null;
  episodeNumber: number | null;
  airDate: string | null;
}

export interface TmdbSeasonInfo {
  seasonNumber: number;
  episodeCount: number | null;
  name?: string | null;
  airDate?: string | null;
}

export interface TmdbEpisodeDetail {
  episodeNumber: number;
  seasonNumber?: number | null;
  name?: string | null;
  overview?: string | null;
  stillPath?: string | null;
  airDate?: string | null;
  runtime?: number | null;
}

export interface TmdbItem {
  id: number;
  mediaType?: string | null;
  title?: string | null;
  name?: string | null;
  numberOfSeasons?: number | null;
  numberOfEpisodes?: number | null;
  seasons?: TmdbSeasonInfo[] | null;
  lastEpisodeToAir?: TmdbEpisodeRef | null;
  nextEpisodeToAir?: TmdbEpisodeRef | null;
}

function ymd(s: string | null | undefined): [number, number, number] | null {
  if (!s || s.length < 10) return null;
  const parts = s.substring(0, 10).split('-');
  if (parts.length !== 3) return null;
  const y = Number(parts[0]);
  const m = Number(parts[1]);
  const d = Number(parts[2]);
  if (Number.isNaN(y) || Number.isNaN(m) || Number.isNaN(d)) return null;
  return [y, m, d];
}

export function isFutureDate(dateStr: string | null | undefined): boolean {
  const parts = ymd(dateStr);
  if (!parts) return false;
  const [y, m, d] = parts;
  const today = new Date();
  today.setHours(0, 0, 0, 0);
  const date = new Date(y, m - 1, d, 0, 0, 0, 0);
  return date.getTime() > today.getTime();
}

export function countEpisodesUpTo(item: TmdbItem, season: number | null, episode: number | null): number {
  if (season == null || episode == null) return item.numberOfEpisodes ?? 0;
  let count = 0;
  for (const s of (item.seasons ?? []).filter((s) => s.seasonNumber !== 0)) {
    if (s.seasonNumber < season) count += s.episodeCount ?? 0;
    else if (s.seasonNumber === season) count += episode;
  }
  return count;
}

export function effectiveLastEpisode(item: TmdbItem): [number, number] | null {
  const nea = item.nextEpisodeToAir;
  if (nea?.seasonNumber != null && nea.episodeNumber != null && nea.airDate != null && !isFutureDate(nea.airDate)) {
    return [nea.seasonNumber, nea.episodeNumber];
  }
  const lea = item.lastEpisodeToAir;
  if (lea?.seasonNumber != null && lea.episodeNumber != null) {
    return [lea.seasonNumber, lea.episodeNumber];
  }
  return null;
}

export function airedEpisodesCount(item: TmdbItem): number {
  const lea = effectiveLastEpisode(item);
  if (!lea) return 0;
  return countEpisodesUpTo(item, lea[0], lea[1]);
}

export function airedEpisodesInSeason(item: TmdbItem, season: number): number {
  const info = (item.seasons ?? []).find((s) => s.seasonNumber === season);
  if (!info) return 0;
  const total = info.episodeCount ?? 0;
  const lea = effectiveLastEpisode(item);
  if (!lea) return total;
  const [lastSeason, lastEp] = lea;
  if (season < lastSeason) return total;
  if (season > lastSeason) return 0;
  return Math.min(total, lastEp);
}

export function episodesBefore(item: TmdbItem, season: number, episode: number): number {
  if (season <= 0) return 0;
  return countEpisodesUpTo(item, season, Math.max(0, episode - 1));
}

export function availableSeasons(item: TmdbItem): number[] {
  const seasons = (item.seasons ?? []).filter((s) => s.seasonNumber > 0);
  let nums: number[];
  const last = effectiveLastEpisode(item);
  if (last) {
    const lastAired = last[0];
    nums = seasons.filter((s) => s.seasonNumber <= lastAired).map((s) => s.seasonNumber);
  } else {
    nums = seasons.filter((s) => s.airDate != null && !isFutureDate(s.airDate)).map((s) => s.seasonNumber);
  }
  nums.sort((a, b) => a - b);
  return nums.length ? nums : [1];
}

export function nextEpisode(item: TmdbItem, season: number, episode: number): [number, number] | null {
  const currentAired = airedEpisodesInSeason(item, season);
  if (currentAired > 0 && episode + 1 <= currentAired) {
    return [season, episode + 1];
  }
  const future = (item.seasons ?? [])
    .filter((s) => s.seasonNumber > season && airedEpisodesInSeason(item, s.seasonNumber) > 0)
    .sort((a, b) => a.seasonNumber - b.seasonNumber)[0];
  return future ? [future.seasonNumber, 1] : null;
}

export function previousEpisode(item: TmdbItem, season: number, episode: number): [number, number] | null {
  if (episode > 1) return [season, episode - 1];
  const past = (item.seasons ?? [])
    .filter((s) => s.seasonNumber < season && airedEpisodesInSeason(item, s.seasonNumber) > 0)
    .sort((a, b) => b.seasonNumber - a.seasonNumber)[0];
  return past ? [past.seasonNumber, airedEpisodesInSeason(item, past.seasonNumber)] : null;
}

export function airedEpisodeList(
  episodes: TmdbEpisodeDetail[],
  item: TmdbItem,
  season: number,
): TmdbEpisodeDetail[] {
  const sorted = [...episodes].sort((a, b) => a.episodeNumber - b.episodeNumber);
  const lea = effectiveLastEpisode(item);
  if (lea) {
    const [lastSeason, lastEp] = lea;
    if (season < lastSeason) return sorted;
    if (season > lastSeason) return [];
    return sorted.filter((e) => e.episodeNumber <= lastEp);
  }
  return sorted.filter((e) => e.airDate != null && !isFutureDate(e.airDate));
}

// ponytail: one runnable self-check. Mirrors the nextEpisode/previousEpisode
// contract: at the last aired episode of S1, next crosses to the next aired
// season; previous of S2E1 lands on the prior season's last aired episode.
export function demo(): void {
  const item: TmdbItem = {
    id: 1,
    numberOfEpisodes: 20,
    seasons: [
      { seasonNumber: 1, episodeCount: 10, airDate: '2020-01-01' },
      { seasonNumber: 2, episodeCount: 10, airDate: '2021-01-01' },
    ],
    lastEpisodeToAir: { seasonNumber: 2, episodeNumber: 5, airDate: '2021-02-01' },
    nextEpisodeToAir: { seasonNumber: 2, episodeNumber: 6, airDate: '2099-01-01' },
  };
  // S2E5 is the last aired → next is S2E6 only if 6 <= airedInSeason(2)=5 → false →
  // no later aired season → null.
  console.assert(nextEpisode(item, 2, 5) === null, 'nextEpisode at last aired should be null');
  // previousEpisode(S2E1) → S1 last aired = 10.
  const prev = previousEpisode(item, 2, 1);
  console.assert(prev !== null && prev[0] === 1 && prev[1] === 10, 'previousEpisode S2E1 -> S1E10');
  // isFutureDate on a clearly future date is true, on a past date false.
  console.assert(isFutureDate('2099-12-31') === true, 'future date');
  console.assert(isFutureDate('2000-01-01') === false, 'past date');
  console.assert(isFutureDate(null) === false, 'null date');
  console.assert(isFutureDate('garbage') === false, 'garbage date');
  console.log('tvlogic.ts demo: OK');
}

// ponytail: run the self-check only under Node (tsx). `typeof process` is safe
// in the browser (where `process` is undefined) so this never throws there.
if (typeof process !== 'undefined' && process.argv[1]?.endsWith('tvlogic.ts')) demo();